plutus-conformance/executables/test-utils/Spec.hs

```diff
@@ -0,0 +1,132 @@
+        "or debug (for debugging failed tests).")
+
+data Runner =
+  Eval
+  | Typecheck
+  deriving stock (Show)
+
+args :: ParserInfo Args
+args = info ((MkArgs <$> ext <*> dir <*> feature) <**> helper)
+  -- using progDescDoc instead of progDesc because progDesc messes up the formatting.
+  (fullDesc <> progDescDoc (Just helpText))
+
+helpText :: Doc
+helpText = string $ unlines
+  ["This program adds test outputs to specified inputs."
+  , "To run the program, input the following 3 arguments:"
+  , "(1) file extension to be searched"
+  , "(2) directory to be searched"
+  , "(3) the action to run the input files through;"
+  , "eval (for evaluation tests),"
+  , "or typecheck (for typechecking tests),"
+  , "or debug (for debugging failed tests)."
+  , "E.g. run \n"
+  , "cabal run test-utils .uplc plutus-conformance/test-cases/uplc/ eval -- --missing \n"
+  , "to have the executable search for files with extension `.uplc`"
+  , "in the /uplc directory."
+  , "It will evaluate and create output files for them."
+  ]
+
+main :: IO ()
+main = do
+    MkArgs extension directory run <- customExecParser (prefs showHelpOnEmpty) args
+    inputFiles <- findByExtension [extension] directory
+    for_ inputFiles $ \inputFile -> do
+      inputTxt <- T.readFile inputFile
+      let parsed = parseTxt inputTxt
+          outFilePath = inputFile <> ".expected"
+      case parsed of
+        Left (ParseErrorB _) ->
+          case run of
+            GenTestOutput Eval ->
+              do -- specifying parsed to ParserError for the compiler
+              -- warn the user that the file failed to parse
+              putStrLn $ inputFile <> " failed to parse. Error written to " <> outFilePath
+              T.writeFile outFilePath shownParseError
+            _ -> pure () -- for debug mode we don't want to see the failed to parse case.
+        Right pro -> do
+          case run of
+            GenTestOutput Eval ->
+              case evalUplcProg (() <$ pro) of
+                (Just prog) -> do
+                  T.writeFile outFilePath (render $ pretty prog)
+                  putStrLn $ inputFile <> " evaluated; result written to " <> outFilePath
+                Nothing      -> do
+                  -- warn the user that the file failed to evaluate
+                  T.writeFile outFilePath shownEvaluationFailure
+                  putStrLn $ inputFile <> " failed to evaluate. Failure written to " <> outFilePath
+            GenTestOutput Typecheck ->
+              putStrLn $
+                "typechecking has not been implemented yet." <>
+                  "Only evaluation tests (eval) or debugging (debug) are supported."
+            Debug ->
+              case agdaEvalUplcProgDebug (() <$ pro) of
+                (Right _prog) -> pure ()
+                (Left err)      ->
+                  -- warn the user that the file failed to evaluate
+                  putStrLn $ inputFile <> " failed to evaluate. " <> err
```

Comment:
Aren't some test cases expected to fail? We shouldn't print the debug information in that case.

Reply:
As mentioned above, the debugger is for those test cases too.